lnwallet/musig_session_test.go

```diff
@@ -0,0 +1,247 @@
+package lnwallet
+
+import (
+	"testing"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/btcec/v2/schnorr/musig2"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightningnetwork/lnd/input"
+	"github.com/lightningnetwork/lnd/keychain"
+	"github.com/stretchr/testify/require"
+)
+
+// nodeType is an enum that represents the two nodes in our test harness.
+type nodeType uint8
+
+const (
+	// nodeAlice is the node that initiates the session.
+	nodeAlice nodeType = iota
+
+	// nodeBob is the node that responds to the session.
+	nodeBob
+)
+
+type muSessionHarness struct {
+	aliceCommit *wire.MsgTx
+	bobCommit   *wire.MsgTx
+
+	aliceSession *MusigPairSession
+
+	bobSession *MusigPairSession
+
+	t *testing.T
+}
+
+func (h *muSessionHarness) selectSession(nodeName nodeType) *MusigPairSession {
+	var targetSession *MusigPairSession
+	switch nodeName {
+	case nodeAlice:
+		targetSession = h.aliceSession
+	case nodeBob:
+		targetSession = h.bobSession
+	}
+
+	return targetSession
+}
+
+func (h *muSessionHarness) refreshSession(nodeName nodeType,
+	nextNonce *musig2.Nonces, revoke bool) {
+
+	var session *MusigPairSession
+	switch nodeName {
+	case nodeAlice:
+		session = h.aliceSession
+	case nodeBob:
+		session = h.bobSession
+	}
+
+	var err error
+
+	// If this isn't in response to a revoke, then we just signed, so we'll
+	// refresh our local session with the newly generated verification
+	// nonce.
+	if !revoke {
+		session.LocalSession, err = session.LocalSession.Refresh(
+			nextNonce,
+		)
+	} else {
+		session.RemoteSession, err = session.RemoteSession.Refresh(
+			nextNonce,
+		)
+	}
+	require.NoError(h.t, err)
+}
+
+// SignCommitment signs a new remote commitment. This is equivalent to sending
+// a CommitSig message on the normal LN protocol.
+func (h *muSessionHarness) SignCommitment(nodeName nodeType) *MusigPartialSig {
+	targetSession := h.selectSession(nodeName)
+
+	sig, err := targetSession.RemoteSession.SignCommit(h.bobCommit)
+	require.NoError(h.t, err)
+
+	return sig
+}
+
+// VerifyAndSignCommitment verifies a remote commitment, then signs a new
+// commitment. This combines receiving a signature, then sending a revoke
+// message.
+func (h *muSessionHarness) VerifyAndSignCommitment(nodeName nodeType,
+	sig *MusigPartialSig) (*MusigPartialSig, *musig2.Nonces) {
+
+	muSession := h.selectSession(nodeName)
+
+	// Verify the commitment transaction from the remote party. The nonce
+	// returned will be sent along side the "revoke and ack" message in the
+	// actual p2p protocol.
+	nextVerificationNonce, err := muSession.LocalSession.VerifyCommitSig(
+		h.aliceCommit, sig.ToWireSig(),
```

Comment:
i think this Verify check currently passes only because `h.aliceCommit` and `h.bobCommit` are identical in this test. Since bob calls this method in order to verify that the sig alice sent him is valid for _his_ local commitment, the input param here should be `h.bobCommit`

Reply:
Yep see comment above, changing the commitment txns here showed the bug (test fails). New version has diff commitment txns, and passes properly.